Transaction index OpCode (#4363)
ethereum/EIPs: eip-4363.md

---
eip: 4363
title: Transaction index OpCode
author: Edgar (@fxfactorial), Dean (@decanus)
status: Draft
type: Standards Track
category: Core
created: 2021-10-12

> ---

Reviewer: It appears you're missing the `description` and `discussions-to` fields of the preamble.

## Abstract

> This EIP adds an opcode that returns the current transactions.

Reviewer: ```suggestion
Adds an opcode that returns the current transaction's index within the block.
```

## Motivation

> Currently users are completely at the mercy of those who order transactions. This means that malicious actors can exploit the ordering of transactions at the users expense. A popular example of this is sandwhich attacks on AMMs, wherein a transaction is put in between 2 others that front-run it.

Reviewer 2: I think it would make sense extending this to get an actual example how this helps. How is the user supposed to influence at which index this should be included? Only by paying a high enough premium so it gets included at the right index, or do you plan some kind of annotation for pools, or specific pools which prefer such transactions?

Reviewer 3: > but I don't see how the end result is any different than priority gas auctions where bots bail out when they lose and cancel their pending transaction

I'd argue that one is a P2P layer solution, the other is an EVM layer solution for ordering, and I'd imagine an EVM layer approach is 100% successful versus late cancelling self-txs. 

And this EIP wouldn't prevent a centralised service (like flashbots) from pruning out reverting transactions to save them hitting the chain.

## Specification

Adds a new opcode `TXINDEX` at 0x49, which uses 0 stack arguments. It pushes the current transaction index onto the stack. Transaction Index is a 256-bit value. The operation costs G_base to execute.

## Reference Implementation

 - A dummy implementation in [go-ethereum@feature/tx-index](https://github.com/dialecticch/go-ethereum/tree/feature/tx-index).
 - Usage of the opcode in [UniswapV3](https://github.com/fxfactorial/uniswap-v3-core/commit/2e605d8f3fe4ccc8c8c0c78eb71674c220348b65).

> 

Reviewer: Also missing Rationale + Security Considerations.
